### PR TITLE
koordlet: tc plugin for netqos

### PR DESCRIPTION
**Pull request by lucming**
### Ⅰ. Describe what this PR does
A plugin for netqos, and based on tc.
other discussion:  https://github.com/koordinator-sh/koordinator/discussions/1764

btw, there may be some duplicate defination with pr https://github.com/koordinator-sh/koordinator/pull/1843 . 

### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

**Comment by zwzhang0107** [comment]
it will be great if you can add some implement details in doc about tc plugin https://github.com/koordinator-sh/koordinator/blob/main/docs/proposals/koordlet/20231208-support-netqos.md#internal-plugins

**Comment by lucming** [comment]
> it will be great if you can add some implement details in doc about tc plugin https://github.com/koordinator-sh/koordinator/blob/main/docs/proposals/koordlet/20231208-support-netqos.md#internal-plugins

write in this pr?

**Comment by zwzhang0107** [comment]
> > it will be great if you can add some implement details in doc about tc plugin https://github.com/koordinator-sh/koordinator/blob/main/docs/proposals/koordlet/20231208-support-netqos.md#internal-plugins
> 
> write in this pr?

another one with only proposal is better

**Comment by lucming** [comment]
> > > it will be great if you can add some implement details in doc about tc plugin https://github.com/koordinator-sh/koordinator/blob/main/docs/proposals/koordlet/20231208-support-netqos.md#internal-plugins
> > 
> > 
> > write in this pr?
> 
> another one with only proposal is better

done

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+
+const (
+	name        = "tcPlugin"
+	description = "setup tc rules for node"
+
+	ruleNameForNodeSLO = name + " (nodeSLO)"
+	ruleNameForAllPods = name + " (allPods)"
+)
+
+const (
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:52]: High -> System
Mid -> LS
Low -> BE

File: pkg/koordlet/runtimehooks/hooks/tc/iptables.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		return fmt.Errorf("can't create tc iptables rulues,because qos manager is nil")
+	}
+	ipsetToClassid := map[extension.NetQoSClass]string{
+		extension.NETQoSHigh: "0001:0002",
+		extension.NETQoSMid:  "0001:0003",
+		extension.NETQoSLow:  "0001:0004",
+	}
+
+	for ipsetName, classid := range ipsetToClassid {
+		exp := fmt.Sprintf("-A POSTROUTING -m set --match-set %s src -j CLASSIFY --set-class %s", string(ipsetName), classid)
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/tc/iptables.go:43]: how to make sure that system process(kubelet) belongs to High(System)

Comment (lucming) [reply pkg/koordlet/runtimehooks/hooks/tc/iptables.go:43]: maybe `1:10` class can be defined as default, package without any marks will be sent to this class by default, WDYT？

**Comment by zwzhang0107** [comment]
image

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+
+	// default policy enables
+	isNETQOSPolicyTC := qosStrategy == nil || qosStrategy.Policies == nil || qosStrategy.Policies.NETQOSPolicy == nil ||
+		*qosStrategy.Policies.NETQOSPolicy == slov1alpha1.NETQOSPolicyTC
+
+	if isNETQOSPolicyTC {
+		if mergedNodeSLO.SystemStrategy == nil {
+			return false, nil
+		}
+		p.rule.enable = true
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:177]: it seems a mutex is needed for rule read&update

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+		klog.Errorf("failed to init some necessary rules. err=%v", err)
+		return err
+	}
+
+	return nil
+}
+
+func (p *tcPlugin) ruleUpdateCb(target *statesinformer.CallbackTarget) error {
+	if !p.initialized.Load() {
+		if err := p.prepare(); err != nil {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:232]: we need to refresh Qdisc once the nodeSLO is updated

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+)
+
+type tcPlugin struct {
+	rule *tcRule
+
+	initialized     *atomic.Bool // whether the cache has been initialized
+	allPodsSyncOnce sync.Once    // sync once for AllPods
+
+	// this is the physical NIC on host, default eth0
+	interfLink netlink.Link
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:100]: need a mutex for interfLink?

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+		rule.WithParseFunc(statesinformer.RegisterTypeAllPods, p.parseForAllPods),
+		rule.WithUpdateCallback(p.ruleUpdateCb))
+
+	reconciler.RegisterCgroupReconciler(reconciler.PodLevel, sysutil.NetClsClassId, description+" (pod net class id)",
+		p.SetPodNetCls, reconciler.PodHostNetworkFilter(), "true")
+
+	p.executor = op.Executor
+}
+
+func (p *tcPlugin) SetPodNetCls(proto protocol.HooksProtocol) error {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:146]: where do we map the cgroup.net_cls to tc.class

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go (hunk 5 above)
Comment (lucming) [reply pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:98]: maybe not.

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go (hunk 6 above)
Comment (lucming) [reply pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:144]: do you have any idea?

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 7]
```diff
@@ -0,0 +1,745 @@
...
+		rule:            newRule(),
+	}
+}
+
+func (p *tcPlugin) Register(op hooks.Options) {
+	klog.V(5).Infof("register hook %v", name)
+
+	rule.Register(ruleNameForNodeSLO, description,
+		rule.WithParseFunc(statesinformer.RegisterTypeNodeSLOSpec, p.parseRuleForNodeSLO),
+		rule.WithUpdateCallback(p.ruleUpdateCb),
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:132]: maybe only need prepare for nodeSLO callback?

Comment (lucming) [reply pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:132]: ipset need to update from pod.

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+	if err := p.InitRelatedRules(); err != nil {
+		klog.Errorf("failed to init some necessary rules. err=%v", err)
+		return err
+	}
+
+	return nil
+}
+
+func (p *tcPlugin) ruleUpdateCb(target *statesinformer.CallbackTarget) error {
+	if err := p.prepare(); err != nil {
```
[the remark this answers is not in the document]
Comment (lucming) [reply pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:271]: yes, i'll fix it.

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 9]
```diff
@@ -0,0 +1,745 @@
...
+}
+
+func (p *tcPlugin) parseForAllPods(e interface{}) (bool, error) {
+	_, ok := e.(*struct{})
+	if !ok {
+		return false, fmt.Errorf("invalid rule type %T", e)
+	}
+
+	needSync := false
+	p.allPodsSyncOnce.Do(func() {
```
[the remark this answers is not in the document]
Comment (lucming) [reply pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:235]: maybe there's a problem here, it wired, i didn't met this when testing. let me take more test.

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 10]
```diff
@@ -0,0 +1,745 @@
...
+			// pod in host network namespace, network bandwidth can be limited by writing to net_cls.classid.
+			continue
+		}
+
+		netqos := GetPodNetQoSClass(pod.Pod)
+		if netqos == NETQoSNone {
+			continue
+		}
+
+		limitInPodLevel, err := p.createRulesForPod(string(pod.Pod.UID), pod.Pod.Annotations, pod.Pod.Status.PodIP, netqos)
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:308]: persist in cgroup file content
map <uid:netcls.id>

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 11]
```diff
@@ -0,0 +1,745 @@
...
+
+		for ip := range ipInIpset[setName].Difference(ipInK8S[setName]) {
+			klog.V(5).Infof("ready to del ip %s from ipset %s.\n", ip, setName)
+			if err := netlink.IpsetDel(setName, &netlink.IPSetEntry{IP: net.ParseIP(ip).To4()}); err != nil {
+				klog.Warningf("failed to write ip %s to ipset %s, err=%v", ip, setName, err)
+				return err
+			}
+		}
+	}
+
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:351]: deleteRulesForPod

**Comment by zwzhang0107** [comment]
go mod tidy?

File: pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go [hunk 12]
```diff
@@ -0,0 +1,953 @@
...
+		return exec.New().Command("tc", "filter", op, "dev", p.interfLink.Attrs().Name,
+			"protocol", "ip", "prio", fmt.Sprintf("%d", prio),
+			"parent", fmt.Sprintf("%d:", ROOT_CLASS_MINOR_ID),
+			"handle", fmt.Sprintf("%d:", minor),
+			"cgroup",
+		)
+	}
+
+	return apierror.NewAggregate([]error{
+		p.ensureFilter(filterKey(strconv.Itoa(SYSTEM_CLASS_MINOR_ID)), genFilterCmd("add", SYSTEM_CLASS_PRIO, SYSTEM_CLASS_MINOR_ID),
```
[the remark this answers is not in the document]
Comment (zwzhang0107) [reply pkg/koordlet/runtimehooks/hooks/tc/tc_linux.go:681]: // TODO

**Comment by zwzhang0107** [comment]
code needs rebase
make lint?

**Pull request merged**